src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+    case CIRCPAD_DIST_LOG_LOGISTIC:
+      p = crypto_rand_double();
+      /* https://en.wikipedia.org/wiki/Log-logistic_distribution#Quantiles
+       * param1 is Alpha, param2 is Beta */
+      return dist.param1 * pow(p/(1.0-p), 1.0/dist.param2);
+    case CIRCPAD_DIST_GEOMETRIC:
+      p = crypto_rand_double();
+      /* https://github.com/distributions-io/geometric-quantile/
+       * param1 is 'p' (success probability) */
+      return ceil(tor_mathlog(1.0-p)/tor_mathlog(1.0-dist.param1));
+    case CIRCPAD_DIST_WEIBULL:
+      p = crypto_rand_double();
+      /* https://en.wikipedia.org/wiki/Weibull_distribution \
+       *    #Cumulative_distribution_function
+       * param1 is k, param2 is Lambda */
+      return dist.param2*pow(-tor_mathlog(1.0-p), 1.0/dist.param1);
+    case CIRCPAD_DIST_PARETO:
+      p = 1.0-crypto_rand_double(); // Pareto quantile needs (0,1]
+
+      /* https://en.wikipedia.org/wiki/Generalized_Pareto_distribution \
+       *    #Generating_generalized_Pareto_random_variables
+       * param1 is Sigma, param2 is Xi
+       * Since it's piecewise, we must define it for 0 (or close to 0) */
+      if (fabs(dist.param2) <= 1e-22)
+        return -dist.param1*tor_mathlog(p);
+      else
+        return dist.param1*(pow(p, -dist.param2) - 1.0)/dist.param2;
+  }
+  return 0;
+}
+
+/**
+ * Find the index of the first bin whose upper bound is
+ * greater than the target, and that has tokens remaining.
+ */
+static int
+circpad_machine_first_higher_index(circpad_machineinfo_t *mi,
+                                   circpad_delay_t target_bin_us)
+{
+  int i = circpad_histogram_usec_to_bin(mi, target_bin_us);
+
+  /* Don't remove from the infinity bin */
+  for (; i < mi->histogram_len-1; i++) {
+    if (mi->histogram[i] &&
+        circpad_histogram_bin_to_usec(mi, i+1) > target_bin_us) {
+      return i;
+    }
+  }
+
+  return mi->histogram_len;
+}
+
+/**
+ * Find the index of the first bin whose lower bound is
+ * lower than the target, and that has tokens remaining.
+ */
+static int
+circpad_machine_first_lower_index(circpad_machineinfo_t *mi,
+                                  circpad_delay_t target_bin_us)
+{
+  int i = circpad_histogram_usec_to_bin(mi, target_bin_us);
+
+  for (; i >= 0; i--) {
+    if (mi->histogram[i] &&
+        circpad_histogram_bin_to_usec(mi, i) <= target_bin_us) {
+      return i;
+    }
+  }
+
+  return -1;
+}
+
+/**
+ * Remove a token from the first non-empty bin whose upper bound is
+ * greater than the target.
+ */
+void
+circpad_machine_remove_higher_token(circpad_machineinfo_t *mi,
+                                    circpad_delay_t target_bin_us)
+{
+  /* We need to remove the token from the first bin
+   * whose upper bound is greater than the target, and that
+   * has tokens remaining. */
+  int i = circpad_machine_first_higher_index(mi, target_bin_us);
+
+  if (i >= 0 && i < mi->histogram_len-1) {
+    tor_assert(mi->histogram[i]);
```

Comment:
This kind of assertion is worrying.  Is this kind of a bug really worth crashing Tor over?  Remember, we've had a lot of DoS bugs in the past caused by logic errors that ran into a tor_assert().

Please go over this file and see how many of the assertions in it could use BUG or tor_assert_nonfatal() instead.

Reply:
I'd suggest a decrement macro that does something like `if (!BUG(foo == 0)) --foo;`

Reply:
Addressed all of them in 0d7d10e3ad. Please review carefully! :)